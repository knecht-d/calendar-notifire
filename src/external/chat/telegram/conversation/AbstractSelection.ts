import Telegraf, { Context } from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";

export abstract class AbstractSelection<T> {
    constructor(
        protected editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
        protected messageId: number,
    ) {}
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    static setActions(_bot: Telegraf<TelegrafContext>, _selectionHandler: (ctx: Context, value: string) => void) {}
    abstract requestInput(buildingObject: T): Promise<(selection: string) => T>;
}
