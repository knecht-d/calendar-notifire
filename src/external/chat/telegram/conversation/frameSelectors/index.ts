export { DaySelection } from "./DaySelection";
export { HourSelection } from "./HourSelection";
export { MinuteSelection } from "./MinuteSelection";
export { MonthSelection } from "./MonthSelection";
