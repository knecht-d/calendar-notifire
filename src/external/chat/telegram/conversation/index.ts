export * from "./FrameConfigBuilder";
export * from "./TriggerConfigBuilder";
