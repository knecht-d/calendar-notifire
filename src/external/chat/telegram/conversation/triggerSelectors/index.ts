export { DayOfMonthSelection } from "./DayOfMonthSelection";
export { DayOfWeekSelection } from "./DayOfWeekSelection";
export { HourSelection } from "./HourSelection";
export { MinuteSelection } from "./MinuteSelection";
export { RecurrenceTypeSelection } from "./RecurrenceTypeSelection";
export { ToHourSelection } from "./ToHourSelection";
