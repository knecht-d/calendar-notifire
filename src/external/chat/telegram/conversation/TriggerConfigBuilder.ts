import Telegraf from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";
import { IGenericRecurrence } from "../../../../gateways";
import { ILogger } from "../../../logging";
import { AbstractSelection } from "./AbstractSelection";
import {
    DayOfMonthSelection,
    DayOfWeekSelection,
    HourSelection,
    MinuteSelection,
    RecurrenceTypeSelection,
    ToHourSelection,
} from "./triggerSelectors";
import { handleButtonPress } from "./utils";

export class TriggerConfigBuilder {
    private static ALL_BUILDER: { [messageId: number]: TriggerConfigBuilder } = {};
    private static LOGGER: ILogger;
    config: IGenericRecurrence;
    currentResolver: (selection: string) => void;

    constructor(
        private messageId: number,
        private editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
    ) {
        this.config = {};
        this.currentResolver = () => {
            return;
        };
        TriggerConfigBuilder.ALL_BUILDER[messageId] = this;
    }

    static init(bot: Telegraf<TelegrafContext>, logger: ILogger) {
        this.LOGGER = logger;
        [
            RecurrenceTypeSelection,
            DayOfMonthSelection,
            DayOfWeekSelection,
            HourSelection,
            ToHourSelection,
            MinuteSelection,
        ].forEach(selection => {
            selection.setActions(bot, this.handleButtonPress.bind(this));
        });
    }

    static handleButtonPress(ctx: TelegrafContext, selection: string) {
        handleButtonPress(this.ALL_BUILDER, this.LOGGER, ctx, selection);
    }

    async requestConfig() {
        await this.requestSelection(RecurrenceTypeSelection);

        if (this.config.type === "m") {
            await this.requestSelection(DayOfMonthSelection);
        } else {
            await this.requestSelection(DayOfWeekSelection);
        }

        await this.requestSelection(HourSelection);

        if (this.config.type === "h") {
            await this.requestSelection(ToHourSelection);
        }

        await this.requestSelection(MinuteSelection);

        delete TriggerConfigBuilder.ALL_BUILDER[this.messageId];

        return this.config;
    }

    getConfig() {
        return this.config;
    }

    private async requestSelection(
        Selection: new (
            editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
            messageId: number,
        ) => AbstractSelection<IGenericRecurrence>,
    ) {
        const selection = new Selection(this.editMessage, this.messageId);
        const selectionHandler = await selection.requestInput(this.config);
        return new Promise<void>(res => {
            this.currentResolver = selection => {
                this.config = selectionHandler(selection);
                res();
            };
        });
    }
}
