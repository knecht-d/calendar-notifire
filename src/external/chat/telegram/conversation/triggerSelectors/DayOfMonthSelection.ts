import Telegraf, { Context, Extra, Markup } from "telegraf";
import { IGenericRecurrence } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";

const prefix = "TRIGGER_DAY_OF_MONTH-";
const values = [...Array(31).keys()].map(value => value + 1);

export class DayOfMonthSelection extends AbstractSelection<IGenericRecurrence> {
    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        values.forEach(value => {
            bot.action(`${prefix}${value}`, ctx => selectionHandler(ctx, `${value}`));
        });
    }

    async requestInput(buildingObject: IGenericRecurrence): Promise<(selection: string) => IGenericRecurrence> {
        await this.editMessage(
            `
An welchem Tag im Monat soll die Erinnerung erfolgen?
Gibt es den Tag nicht, dann findet in dem Monat keine Erinnerung statt!
z.B. "30. Februar"`.trim(),
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    values.map(value => Markup.callbackButton(`${value}`, `${prefix}${value}`)),
                    { columns: 4 },
                ),
            ),
        );
        return (selection: string) => {
            buildingObject.dayOfMonth = Number.parseInt(selection);
            return buildingObject;
        };
    }
}
