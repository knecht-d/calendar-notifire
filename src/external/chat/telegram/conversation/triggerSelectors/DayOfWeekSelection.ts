import Telegraf, { Context, Extra, Markup } from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";
import { IGenericRecurrence } from "../../../../../gateways";
import { AbstractSelection } from "../AbstractSelection";

const daysOfWeekPrefix = "TRIGGER_DAYS_OF_WEEK-";
const days = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"];
const done = "DONE";

// const dayOfWeekMonday = `${daysOfWeekPrefix}Mon`;
// const dayOfWeekTuesday = `${daysOfWeekPrefix}Tue`;
// const dayOfWeekWednesday = `${daysOfWeekPrefix}Wed`;
// const dayOfWeekThursday = `${daysOfWeekPrefix}Thu`;
// const dayOfWeekFriday = `${daysOfWeekPrefix}Fri`;
// const dayOfWeekSatuarday = `${daysOfWeekPrefix}Sat`;
// const dayOfWeekSunday = `${daysOfWeekPrefix}Sun`;
// const dayOfWeekDone = `${daysOfWeekPrefix}DONE`;

export class DayOfWeekSelection extends AbstractSelection<IGenericRecurrence> {
    static ALL_SELECTIONS: { [messageId: number]: DayOfWeekSelection } = {};
    daysOfWeek = {
        monday: true,
        tuesday: true,
        wednesday: true,
        thursday: true,
        friday: true,
        saturday: true,
        sunday: true,
    };

    constructor(editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>, messageId: number) {
        super(editMessage, messageId);
        DayOfWeekSelection.ALL_SELECTIONS[this.messageId] = this;
    }

    static setActions(bot: Telegraf<Context>, selectionHandler: (ctx: Context, value: string) => void): void {
        days.forEach(day => {
            bot.action(`${daysOfWeekPrefix}${day}`, ctx => this.handleDay(ctx, day));
        });
        bot.action(`${daysOfWeekPrefix}${done}`, ctx => selectionHandler(ctx, done));
    }

    static handleDay(ctx: TelegrafContext, selection: string) {
        const currentSelection = this.ALL_SELECTIONS[ctx.callbackQuery?.message?.message_id || 0];
        if (currentSelection) {
            const typedSelection = selection as keyof typeof currentSelection.daysOfWeek;
            currentSelection.daysOfWeek[typedSelection] = !currentSelection.daysOfWeek[typedSelection];

            currentSelection.replyForDaysOfWeek().catch(() => {
                // this.LOGGER.error("TriggerConfigBuilder", error)
                currentSelection
                    .editMessage(
                        "Nachricht konnte nicht weiter verarbeitet werden. Versuchen Sie es mit einem erneuten Befehl.",
                    )
                    .catch(() => {
                        // this.LOGGER.error("TriggerConfigBuilder", error)
                    });
            });
        } else {
            ctx.editMessageText(
                "Nachricht konnte nicht weiter verarbeitet werden. Versuchen Sie es mit einem erneuten Befehl.",
            ).catch(() => {
                // this.LOGGER.error("TriggerConfigBuilder", error)
            });
        }
    }

    async requestInput(buildingObject: IGenericRecurrence): Promise<(selection: string) => IGenericRecurrence> {
        await this.replyForDaysOfWeek();
        return () => {
            delete DayOfWeekSelection.ALL_SELECTIONS[this.messageId];
            buildingObject.daysOfWeek = this.daysOfWeek;
            return buildingObject;
        };
    }

    private replyForDaysOfWeek() {
        const mondayBtn = Markup.callbackButton(
            `${this.daysOfWeek.monday ? "✓" : " "}Montag `,
            `${daysOfWeekPrefix}${days[0]}`,
        );
        const tuesdayBtn = Markup.callbackButton(
            `${this.daysOfWeek.tuesday ? "✓" : " "}Dienstag `,
            `${daysOfWeekPrefix}${days[1]}`,
        );
        const wednesdayBtn = Markup.callbackButton(
            `${this.daysOfWeek.wednesday ? "✓" : " "}Mittwoch `,
            `${daysOfWeekPrefix}${days[2]}`,
        );
        const thursdayBtn = Markup.callbackButton(
            `${this.daysOfWeek.thursday ? "✓" : " "}Donnerstag `,
            `${daysOfWeekPrefix}${days[3]}`,
        );
        const fridayBtn = Markup.callbackButton(
            `${this.daysOfWeek.friday ? "✓" : " "}Freitag `,
            `${daysOfWeekPrefix}${days[4]}`,
        );
        const saturdayBtn = Markup.callbackButton(
            `${this.daysOfWeek.saturday ? "✓" : " "}Samstag `,
            `${daysOfWeekPrefix}${days[5]}`,
        );
        const sundayBtn = Markup.callbackButton(
            `${this.daysOfWeek.sunday ? "✓" : " "}Sonntag `,
            `${daysOfWeekPrefix}${days[6]}`,
        );
        const doneBtn = Markup.callbackButton("Fertig", `${daysOfWeekPrefix}${done}`);
        return this.editMessage(
            "An welchen Tagen soll eine Erinnerung erfolgen?",
            Extra.HTML().markup(() =>
                Markup.inlineKeyboard(
                    [
                        [mondayBtn, fridayBtn],
                        [tuesdayBtn, saturdayBtn],
                        [wednesdayBtn, sundayBtn],
                        [thursdayBtn, doneBtn],
                    ],
                    {},
                ),
            ),
        );
    }
}
