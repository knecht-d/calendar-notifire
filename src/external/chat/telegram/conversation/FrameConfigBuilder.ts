import Telegraf from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { ExtraReplyMessage } from "telegraf/typings/telegram-types";
import { IGenericFrame, IGenericRecurrence } from "../../../../gateways";
import { ILogger } from "../../../logging";
import { AbstractSelection } from "./AbstractSelection";
import { DaySelection, HourSelection, MinuteSelection, MonthSelection } from "./frameSelectors";
import { handleButtonPress } from "./utils";

export class FrameConfigBuilder {
    private static ALL_BUILDER: { [messageId: number]: FrameConfigBuilder } = {};
    private static LOGGER: ILogger;
    config: IGenericFrame;
    currentResolver: (selection: string) => void;

    constructor(
        private messageId: number,
        private editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
    ) {
        this.config = {};
        this.currentResolver = () => {
            return;
        };
        FrameConfigBuilder.ALL_BUILDER[messageId] = this;
    }

    static init(bot: Telegraf<TelegrafContext>, logger: ILogger) {
        this.LOGGER = logger;
        [MonthSelection, DaySelection, MinuteSelection, HourSelection].forEach(selection => {
            selection.setActions(bot, this.handleButtonPress.bind(this));
        });
    }

    static handleButtonPress(ctx: TelegrafContext, selection: string) {
        handleButtonPress(this.ALL_BUILDER, this.LOGGER, ctx, selection);
    }

    async requestConfig(recurrence: IGenericRecurrence) {
        switch (recurrence.type) {
            case "m":
                await this.requestSelection(MonthSelection);
                await this.requestSelection(DaySelection);
                await this.requestSelection(HourSelection);
                break;
            case "d":
                await this.requestSelection(DaySelection);
                await this.requestSelection(HourSelection);
                await this.requestSelection(MinuteSelection);
                break;
            case "h":
                await this.requestSelection(HourSelection);
                await this.requestSelection(MinuteSelection);
                break;
            default:
                throw new Error("Unexpected Recurrence Type");
        }

        delete FrameConfigBuilder.ALL_BUILDER[this.messageId];
        return this.config;
    }

    getConfig() {
        return this.config;
    }

    private async requestSelection(
        Selection: new (
            editMessage: (text: string, extra?: ExtraReplyMessage) => Promise<void>,
            messageId: number,
        ) => AbstractSelection<IGenericFrame>,
    ) {
        const selection = new Selection(this.editMessage, this.messageId);
        const selectionHandler = await selection.requestInput(this.config);
        return new Promise<void>(res => {
            this.currentResolver = selection => {
                this.config = selectionHandler(selection);
                res();
            };
        });
    }
}
