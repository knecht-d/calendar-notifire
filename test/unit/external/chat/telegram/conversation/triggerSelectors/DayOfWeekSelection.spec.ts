import { Extra, Markup } from "telegraf";
import { TelegrafContext } from "telegraf/typings/context";
import { DayOfWeekSelection } from "../../../../../../../src/external/chat/telegram/conversation/triggerSelectors";

describe("DayOfWeekSelection", () => {
    const days = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"];
    const done = "DONE";
    const MockBot = {
        action: jest.fn((_, _handler) => {
            // handler(mockCtx);
        }),
    };
    const mockEditMessage = jest.fn(() => Promise.resolve());

    beforeEach(() => {
        jest.clearAllMocks();
        jest.spyOn(Extra, "markup").mockImplementation((): any => {
            return {
                markup: (cb: () => any) => cb(),
            };
        });
        jest.spyOn(Markup, "callbackButton").mockImplementation((desc: string, key: string): any => `${desc}(${key})`);
        jest.spyOn(Markup, "inlineKeyboard");
    });

    describe("setActions", () => {
        it.each([...days, done])("should set action for %s", value => {
            const mockSelectionHandler = jest.fn();
            DayOfWeekSelection.setActions(MockBot as any, mockSelectionHandler);
            expect(MockBot.action).toHaveBeenCalledWith(`TRIGGER_DAYS_OF_WEEK-${value}`, expect.any(Function));
        });
    });

    describe("handler", () => {
        const mockCtx = {
            editMessageText: jest.fn(),
            // eslint-disable-next-line @typescript-eslint/camelcase
            callbackQuery: { message: { message_id: 42 } },
        };
        const handlers: { [key: string]: (ctx: TelegrafContext) => void } = {};
        let selection: DayOfWeekSelection;
        beforeEach(() => {
            MockBot.action.mockImplementation((key: string, handler: (ctx: TelegrafContext) => void) => {
                handlers[key] = handler;
            });
            const mockSelectionHandler = jest.fn();
            DayOfWeekSelection.setActions(MockBot as any, mockSelectionHandler);
        });

        it.each([
            "TRIGGER_DAYS_OF_WEEK-monday",
            "TRIGGER_DAYS_OF_WEEK-tuesday",
            "TRIGGER_DAYS_OF_WEEK-wednesday",
            "TRIGGER_DAYS_OF_WEEK-thursday",
            "TRIGGER_DAYS_OF_WEEK-friday",
            "TRIGGER_DAYS_OF_WEEK-saturday",
            "TRIGGER_DAYS_OF_WEEK-sunday",
        ])("should handle %s", triggerKey => {
            selection = new DayOfWeekSelection(mockEditMessage, 42);
            handlers[triggerKey](mockCtx as any);

            expect((selection as any).daysOfWeek).toMatchSnapshot("daysofweek");
            expect((Markup.inlineKeyboard as jest.Mock).mock.calls).toMatchSnapshot("keyboard");
        });
    });

    describe("requestInput", () => {
        it("should send the question for type Hourly", async () => {
            const selection = new DayOfWeekSelection(mockEditMessage, 42);
            await selection.requestInput({});
            expect(mockEditMessage).toHaveBeenCalledWith(
                expect.stringContaining("An welchen Tagen"),
                expect.any(Object),
            );
        });

        it("should have the correct keyboard", async () => {
            const selection = new DayOfWeekSelection(mockEditMessage, 42);
            await selection.requestInput({});

            expect((Markup.inlineKeyboard as jest.Mock).mock.calls).toMatchSnapshot();
        });

        it("should return the days on selection of DONE", async () => {
            const selection = new DayOfWeekSelection(mockEditMessage, 42);
            const handler = await selection.requestInput({});
            (selection as any).daysOfWeek = {
                monday: true,
                tuesday: false,
                wednesday: true,
                thursday: false,
                friday: true,
                saturday: false,
                sunday: true,
            };
            const result = handler("");
            expect(result).toEqual({
                daysOfWeek: {
                    monday: true,
                    tuesday: false,
                    wednesday: true,
                    thursday: false,
                    friday: true,
                    saturday: false,
                    sunday: true,
                },
            });
        });
    });
});
