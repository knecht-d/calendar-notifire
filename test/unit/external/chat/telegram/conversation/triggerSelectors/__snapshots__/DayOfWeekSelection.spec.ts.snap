// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-friday: daysofweek 1`] = `
Object {
  "friday": false,
  "monday": true,
  "saturday": true,
  "sunday": true,
  "thursday": true,
  "tuesday": true,
  "wednesday": true,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-friday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        " Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-monday: daysofweek 1`] = `
Object {
  "friday": true,
  "monday": false,
  "saturday": true,
  "sunday": true,
  "thursday": true,
  "tuesday": true,
  "wednesday": true,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-monday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        " Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-saturday: daysofweek 1`] = `
Object {
  "friday": true,
  "monday": true,
  "saturday": false,
  "sunday": true,
  "thursday": true,
  "tuesday": true,
  "wednesday": true,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-saturday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        " Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-sunday: daysofweek 1`] = `
Object {
  "friday": true,
  "monday": true,
  "saturday": true,
  "sunday": false,
  "thursday": true,
  "tuesday": true,
  "wednesday": true,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-sunday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        " Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-thursday: daysofweek 1`] = `
Object {
  "friday": true,
  "monday": true,
  "saturday": true,
  "sunday": true,
  "thursday": false,
  "tuesday": true,
  "wednesday": true,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-thursday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        " Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-tuesday: daysofweek 1`] = `
Object {
  "friday": true,
  "monday": true,
  "saturday": true,
  "sunday": true,
  "thursday": true,
  "tuesday": false,
  "wednesday": true,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-tuesday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        " Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-wednesday: daysofweek 1`] = `
Object {
  "friday": true,
  "monday": true,
  "saturday": true,
  "sunday": true,
  "thursday": true,
  "tuesday": true,
  "wednesday": false,
}
`;

exports[`DayOfWeekSelection handler should handle TRIGGER_DAYS_OF_WEEK-wednesday: keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        " Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;

exports[`DayOfWeekSelection requestInput should have the correct keyboard 1`] = `
Array [
  Array [
    Array [
      Array [
        "✓Montag (TRIGGER_DAYS_OF_WEEK-monday)",
        "✓Freitag (TRIGGER_DAYS_OF_WEEK-friday)",
      ],
      Array [
        "✓Dienstag (TRIGGER_DAYS_OF_WEEK-tuesday)",
        "✓Samstag (TRIGGER_DAYS_OF_WEEK-saturday)",
      ],
      Array [
        "✓Mittwoch (TRIGGER_DAYS_OF_WEEK-wednesday)",
        "✓Sonntag (TRIGGER_DAYS_OF_WEEK-sunday)",
      ],
      Array [
        "✓Donnerstag (TRIGGER_DAYS_OF_WEEK-thursday)",
        "Fertig(TRIGGER_DAYS_OF_WEEK-DONE)",
      ],
    ],
    Object {},
  ],
]
`;
